src/main/java/bio/terra/workspace/service/datareference/utils/DataReferenceValidationUtils.java

```diff
@@ -63,9 +84,55 @@ private void validateSnapshotReference(SnapshotReference ref, AuthenticatedUserR
               + "instanceName and snapshot must both be provided.");
     }
     if (!dataRepoService.snapshotExists(ref.instanceName(), ref.snapshot(), userReq)) {
+      logger.warn(
+          String.format(
+              "Snapshot %s not found in data repo instance %s",
+              ref.snapshot(), ref.instanceName()));
       throw new InvalidDataReferenceException(
           "The given snapshot could not be found in the Data Repo instance provided."
               + " Verify that your reference was correctly defined and the instance is correct");
     }
   }
+
+  /**
+   * Validates that a bucket exists and the user has GET access to it. This makes no other
+   * assumptions about the bucket.
+   */
+  private void validateGoogleBucket(GoogleBucketReference ref, AuthenticatedUserRequest userReq) {
+    try {
+      // StorageCow.get() returns null if the bucket does not exist or a user does not have access,
+      // which fails validation.
+      BucketCow bucket = crlService.createStorageCow(userReq).get(ref.bucketName());
+      if (bucket == null) {
+        logger.warn(String.format("Bucket %s not found", ref.bucketName()));
+        throw new InvalidDataReferenceException(
+            "Could not access specified GCS bucket. Ensure the name is correct and that you have access.");
+      }
+    } catch (StorageException e) {
+      throw new InvalidDataReferenceException("Error while trying to access GCS bucket", e);
+    }
+  }
+
+  /**
+   * Validates that a BQ dataset exists and the user has GET access to it. This makes no other
+   * assumptions about the dataset.
+   */
+  private void validateBigQueryDataset(
+      BigQueryDatasetReference ref, AuthenticatedUserRequest userReq) {
+    try {
+      DatasetId datasetId = DatasetId.of(ref.projectId(), ref.datasetName());
+      // BigQueryCow.get() returns null if the bucket does not exist or a user does not have access,
+      // which fails validation.
+      DatasetCow dataset = crlService.createBigQueryCow(userReq).getDataset(datasetId);
+      if (dataset == null) {
+        logger.warn(
+            String.format(
+                "Dataset %s not found in project %s", ref.datasetName(), ref.projectId()));
+        throw new InvalidDataReferenceException(
+            "Could not access specified BigQuery dataset. Ensure the name and GCP project are correct and that you have access.");
+      }
+    } catch (BigQueryException e) {
+      throw new InvalidDataReferenceException("Error while trying to access BigQuery dataset", e);
+    }
+  }
```

Comment:
I wonder if each `ReferenceObject`  subclass should define their own `validate()` method inside their own class  - similar to how `toJson()` is defined. Then we'd just delegate to their own implementations. This might require passing a `CrlService` or `DataRepoService` to it.

Reply:
Having to pass the services to the objects was the reason I didn't want to take that approach - it feels like we're drifting further away from these being value classes. (They already kind of weren't with things like `toApiModel()`, but adding more members makes serialization/deserialization trickier).

Reply:
What is the benefit of them being value classes if we have to write switch statements to dispatch methods on them?

I think it would be a good separation of concerns to have the reference objects encapsulate the knowledge of the cloud resource they reference.

Reply:
The value is that they can be serialized into the database and back out, which is how we store the actual references. If we require the reference objects to have a `DataRepoService` field, that gets more complicated - do we serialize the `DataRepoService` field with the reference? and if not, how do you deserialize into a valid reference object?

Reply:
I think there is a middle ground. I wrote a doc for further discussion: [Workspace Manager Object Hierarchy](https://docs.google.com/document/d/1yOW5oYeNctijQghEA4PWFMHxHc4p5I5EpUMojDZq14Q/edit#heading=h.x6w7yp6besie)